### ISSUE TITLE
[webpack5]: transpiling fails for package with esm-only package exports

### ISSUE THREAD

**Issue by stefanprobst** [martpie/next-transpile-modules#181]
- [x] I HAVE READ [THE FAQ](https://github.com/martpie/next-transpile-modules#faq) AND MY PROBLEM WAS NOT DESCRIBED THERE
- [x] I WILL GIVE 10$ TO CHARITY IF MY SOLUTION WAS ACTUALLY IN THE README/FAQ

**Are you trying to transpile a local package or an npm package?**

I'm trying to transpile [`xdm`](https://github.com/wooorm/xdm#install), which is an ESM package.

**Describe the bug**

Transpilation works with webpack 4, but failes when enabling webpack5 with `{ future: { webpack5: true } }`.

**To Reproduce**

Clone and run [this repro](https://github.com/stefanprobst/issue-next-webpack5-xdm) which is a default `create-next-app` with added `next.config.js`. Open `http://localhost:3000`. Error log:

```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main defined in /data/Development/playground/issue-next-webpack5-xdm/node_modules/periscopic/package.json
```

The error message points to `periscopic`, which is a dependency of `xdm`. It has [an `exports` field defined in its `package.json`](https://github.com/Rich-Harris/periscopic/blob/master/package.json) which only includes [the "import" conditional export](https://nodejs.org/api/packages.html#packages_conditional_exports).

My guess is that webpack4 will just use the `module` field in package.json, but in webpack5 package exports take precedence, and this breaks stuff. (I realize that the `next-transpile-module` docs include a note to "make sure all your packages have a valid main field.", but i *think* `exports` will always win over `main`/`module`)

When I include `periscopic` in the list of modules to transpile, the build fails early with:

```
Error: next-transpile-modules - an unexpected error happened when trying to resolve "periscopic"
Error: Package path . is not exported from package /data/Development/playground/issue-next-webpack5-xdm/node_modules/periscopic (see exports field in /data/Development/playground/issue-next-webpack5-xdm/node_modules/periscopic/package.json)
```

When adding "import" to the list of `enhanced-resolve`'s `conditionNames` [here](https://github.com/martpie/next-transpile-modules/blob/fa1fdfe60cdb8e5b3945bf361c47c43d1237886c/src/next-transpile-modules.js#L100) things seem to work (but not sure this is the correct way to solve this).

**Expected behavior**

Ideally, Next.js would just support esm. But until then - transpiling `xdm` should work without error.

**Setup**

- Next.js version: 10.1.3
- `next-transpile-modules` version: 6.4.0
- Node.js version: 14.16.0
- `npm`/`yarn` version: yarn 1.22.5
- Operating System: Ubuntu 20.04
- Webpack 4 or 5: 5

**Comment by martpie** [comment]
Usually, the way to solve this is to ask the person responsible of the package to add a `default` export field.

cf. https://github.com/martpie/next-transpile-modules/issues/147 and the linked issues in `nanoid` and `enhanced-resolve`.

I really need to do more research on it. One other way to fix it would be to update the `conditionName`, but it feels weird and like there will always be one case that won't work as expected.

**Comment by stefanprobst** [comment]
thanks for the quick reply!

i also tried to read up on this, and it's all a bit confusing. while the webpack maintainer comment linked to in the source [here](https://github.com/martpie/next-transpile-modules/blob/fa1fdfe60cdb8e5b3945bf361c47c43d1237886c/src/next-transpile-modules.js#L99) seems to suggest to add a `default` export, the webpack 5 docs on conditional exports actually have as the first point in their "Guidelines" to ["Avoid the `default` export. It's handled differently between tooling. Only use named exports."](https://webpack.js.org/guides/package-exports/#guidelines)

I'm also curious if there would be a difference between packages adding a `default` export, and this package adding "import" to `conditionNames`? as i understand it (and that might very well be wrong), in the first case `enhanced-resolve` would check `exports.require` and then `exports.default` to get the package root, and in the second case `exports.require` then `exports.import` and then `exports.default` - which would amount to the same location if upstream aliased "import" to "default"?

one potential issue with both require and import in `conditionNames` is mentioned [here](https://nodejs.org/api/packages.html#packages_dual_package_hazard) and [here](https://esbuild.github.io/api/#how-conditions-work) - but i have no idea if this applies what `next-transpile-modules` does?

**Comment by lunelson** [comment]
@stefanprobst I've been working with `xdm` too. It worked for me with this plugin, up to version `1.7.1` or so, but then the issue with `periscopic` appeared. The particular`exports` problem can be solved with the `conditionNames` parameter on `enhanced-resolve` (see below); but...

```js
const resolve = enhancedResolve.create.sync({
  conditionNames: ['require', 'import', 'node', 'esm'],
});
```

...then somehow `xdm` itself is still resolving the wrong thing and you get the warning about ESM modules again.

My new solution is to [transpile the entire `xdm` library and all its deps with `esbuild`](https://esbuild.github.io/getting-started/#bundling-for-node) to a node v10-compatible bundle.

```sh
yarn add -D esbuild
touch lib/xdm-exports.js
touch lib/xdm-bundled.js
```

Then in `xdm-exports.js` you simply write:

```js
export * from 'xdm';
```

Then bundle:

```sh
esbuild lib/xdm-exports.js --bundle --outfile=lib/xdm-bundled.js --platform=node --target=node10
```

Then you can `require` xdm's exports from `xdm-bundled.js`:

```js
const { compile, compileSync, evaluate, evaluateSync } = require('./xdm-bundled');
```

**Comment by martpie** [comment]
May have been fixed in 7.1.0

**Comment by stefanprobst** [comment]
> May have been fixed in 7.1.0

I'm afraid not:

```
next-transpile-modules - DEPRECATED - fallbacking to previous module resolution system for module "periscopic", you can now just pass the name of the package to transpile and it will detect its real path without you having to pass a sub-module.
Error: next-transpile-modules - an unexpected error happened when trying to resolve "periscopic". Are you sure the name module you are trying to transpile is correct, and it has a "main" or an "exports" field?
Error: Package path . is not exported from package /data/Development/playground/issue-next-webpack5-xdm/node_modules/periscopic (see exports field in /data/Development/playground/issue-next-webpack5-xdm/node_modules/periscopic/package.json)
    at getPackageRootDirectory (/data/Development/playground/issue-next-webpack5-xdm/node_modules/next-transpile-modules/src/next-transpile-modules.js:147:17)
```

**Comment by martpie** [comment]
Can you paste here your NTM initializer so I can at how it looks like?

**Comment by stefanprobst** [comment]
you'll find the `next.config.js` [here](https://github.com/stefanprobst/issue-next-webpack5-xdm/blob/main/next.config.js)

and `periscopic`'s `package.json` is [here](https://github.com/Rich-Harris/periscopic/blob/master/package.json)

**Comment by martpie** [comment]
Ok, that's really weird

```
next-transpile-modules - DEPRECATED - fallbacking to previous module resolution system for module "periscopic", you can now just pass the name of the package to transpile and it will detect its real path without you having to pass a sub-module.
```

should not be happening as your pointing it to `periscopic` directly 🤔  I need to investigate a bit more

**Comment by martpie** [comment]
Ok, so this was partly fixed by 7.1.0, but there is a new problem I did not foresee (exports take the precedence and `require.resolve('periscopic/package.json')` cannot be resolved. I'll fix it soon.

**Comment by martpie** [comment]
Should be fixed in 7.1.1, thank you for the feedback 🙌

**Comment by stefanprobst** [comment]
thanks! :tada:
